spatialos-sdk/src/worker/schema.rs

```diff
@@ -1,47 +1,59 @@
-//! Rust wrappers around the various schema data types defined by the C SDK.
-//!
-//! TODO: Document the conventions around how we handle correctness and safety
-//! around schema data:
-//!
-//! * By convention there are two lifetimes for each type: `'owner` and `'data`. The
-//!   immutable view only uses `'owner`, the owned version uses `'data`.
-//! * There are two Rust types for each C type: An immutable, borrowed version; And
-//!   a mutable, owned version.
-
-use crate::worker::{internal::Sealed, EntityId};
+use crate::worker::{internal::SchemaFieldSealed, EntityId};
 use spatialos_sdk_sys::worker::*;
 use std::{collections::BTreeMap, mem, slice};
 
 mod component_data;
+mod component_update;
 mod object;
 pub mod owned;
 
-pub use self::{component_data::*, object::*};
+pub use self::{component_data::*, component_update::*, object::*};
 
 pub type FieldId = u32;
 
 // =================================================================================================
 // Schema Conversion Traits
 // =================================================================================================
 
-pub trait SchemaField: Sized + Sealed {
+pub trait SchemaField: Sized + SchemaFieldSealed {
     type RustType: Sized;
 
+    fn get_field(object: &Object, field: FieldId) -> Self::RustType;
+
     fn add_field(object: &mut Object, field: FieldId, value: &Self::RustType);
 
-    fn get_field(object: &Object, field: FieldId) -> Self::RustType;
+    fn has_update(update: &ComponentUpdate, field: FieldId) -> bool;
+
+    fn get_update(update: &ComponentUpdate, field: FieldId) -> Option<Self::RustType> {
+        if Self::has_update(update, field) {
+            Some(Self::get_field(update.fields(), field))
+        } else {
+            None
+        }
+    }
+
+    fn add_update(update: &mut ComponentUpdate, field: FieldId, value: &Self::RustType) {
+        Self::add_field(update.fields_mut(), field, value);
+    }
 }
 
 pub trait IndexedField: SchemaField {
     fn field_count(object: &Object, field: FieldId) -> u32;
-
     fn index_field(object: &Object, field: FieldId, index: u32) -> Self::RustType;
 }
 
+/// A type that can be efficiently added to a schema object as a list.
+///
+/// Since types that are serialized/deserialized as a list are copied directly into
+/// the schema object, any types that implement this trait must be ABI-compatible
+/// with the underlying C API. This means that only primitive types (including
+/// [`EntityId`]) implement `ArrayField`.
+///
+/// [`EntityId`]
```

Comment:
Is this line of the docs intended?

Reply:
Ah, my mistake! I meant to make that a link to the relevant docs page, but must have gotten distracted 😅